### PR TITLE
Hash prefixes

### LINKED ISSUE #26: Hash collision
Currently group names and urls are hashed and stored as keys at the "root" of the Redis DB. So if a group's name is an url (strange but possible), this will result in some shady behavior - yet it does not crash.

Best practice would be to prepend hashes with a namespace, e.g. `url:<hash>`, `group:<hash>`...

### PR DESCRIPTION
**Pull request by abulte**
Fix #26.

The migration still has to be tested against a production DB.

### CODE REVIEW

File: croquemort/migrations.py [hunk 1]
```diff
@@ -28,3 +29,59 @@ def split_content_types(self):
...
+                database.rename(url_hash, new_hash)
+                if 'group' in data:
+                    new_g_hash = '{}{}'.format(group_prefix, data['group'])
+                    database.hset(new_hash, 'group', new_g_hash)
+            idx += 1
+        # migrate group hashes and nested url hashes
+        _, all_keys = database.scan()
+        for key in all_keys:
+            if database.type(key) == 'hash' \
+                    and not key.startswith('cache') \
```
Comment (davidbgk) [inline croquemort/migrations.py]: What if you test `':' in key`?

File: croquemort/migrations.py [hunk 2]
```diff
@@ -28,3 +29,59 @@ def split_content_types(self):
...
+        # in url hash
+        for url_hash, data in self.storage.get_all_urls():
+            if not url_hash.startswith(url_prefix):
+                new_hash = '{}{}'.format(url_prefix, url_hash)
+                database.lset('urls', idx, new_hash)
+                database.rename(url_hash, new_hash)
+                if 'group' in data:
+                    new_g_hash = '{}{}'.format(group_prefix, data['group'])
+                    database.hset(new_hash, 'group', new_g_hash)
+            idx += 1
```
Comment (davidbgk) [inline croquemort/migrations.py]: Why no enumerate here?

File: croquemort/migrations.py (hunk 1 above)
Comment (davidbgk) [inline croquemort/migrations.py]: About that specific `cache` prefix, I can't remember where we use it?

Comment (abulte) [inline croquemort/migrations.py]: It's for a cache decorator used in views. I should probably migrate it too, come to think of it.

File: croquemort/migrations.py (hunk 2 above)
Comment (abulte) [inline croquemort/migrations.py]: I knew it ;-) Let me try.

File: croquemort/migrations.py (hunk 1 above)
Comment (abulte) [inline croquemort/migrations.py]: Come to think of it, again, I will let the `cache-` and `check-` keys as is. It's not worth the trouble, they're already prefixed anyway. And they expire, so I guess it's OK (and even beneficial) if their naming scheme is different.

Comment (abulte) [inline croquemort/migrations.py]: But I already migrated the checks... So I should probably migrate cache too for consistency (sorry for the noise).

Comment (abulte) [inline croquemort/migrations.py]: But cache keys are not hashes! Sorry, last time.

**Comment by abulte** [comment]
I ended up refactoring the whole thing because it was not working properly against the production database — I had some strange behavior while iterating on the `urls` hashes, some hashes being flagged as not migrated in the list but not migrated in the DB.

The last version has been tested against the prod DB, it took 7 minutes and looks OK. Difficult to be sure though, especially because `!` is not supported in Redis pattern matching, so I can not look for keys without prefix.

**Comment by abulte** [comment]
It looks like there are some duplicates in the `urls` list, that would explain my previous problem :-/

```
{'u:616e821a', 'u:3a993ade', 'u:55796f5f', 'u:fc1ba37c', 'u:c62b67f6', 'u:8a35a876', 'u:3da4a871', 'u:d9281ccf', 'u:ff1c4c00', 'u:436f7728', 'u:fc8ee4ea', 'u:c29edd49', 'u:e9b510bd', 'u:33a751ea', 'u:6a87bd46', 'u:fd08a7d1', 'u:3a9ac921', 'u:0b7bcaf2', 'u:4dce09d1', 'u:2fb044f3', 'u:b71afb68', 'u:6cb90e11', 'u:b16324ff', 'u:4f10105a', 'u:42f2b10c', 'u:87fe818a', 'u:8cca07eb', 'u:998035f4', 'u:40d991ee', 'u:8741b28a', 'u:14c065e3', 'u:77f7f7a5', 'u:e86a522e', 'u:f85ecafe'}
```

File: croquemort/migrations.py [hunk 3]
```diff
@@ -28,3 +29,87 @@ def split_content_types(self):
...
+                    and not key.startswith('check-') \
+                    and not key.startswith(url_prefix) \
+                    and not key.startswith(group_prefix):
+                data = database.hgetall(key)
+                if data.get('name'):
+                    self._migrate_group(key)
+                elif data.get('url'):
+                    self._migrate_url(key, data)
+                else:
+                    log('/!\ unknow hash type at key {}'.format(key))
```
Comment (davidbgk) [inline croquemort/migrations.py]: unknown*

**Comment by abulte** [comment]
@davidbgk yes, I'll start a PR w/ the 2.0 "meta" stuff.

**Pull request merged**